src/ClientImpl.cpp

```diff
@@ -346,42 +466,31 @@ ClientImpl::ClientImpl(ClientConfig config) throw(voltdb::Exception, voltdb::Lib
 #endif
     m_cfg = event_config_new();
     if (m_cfg == NULL) {
-        throw voltdb::LibEventException();
+        throw LibEventException("Failed to create configuration for event");
     }
     event_config_set_flag(m_cfg, EVENT_BASE_FLAG_NO_CACHE_TIME);//, EVENT_BASE_FLAG_NOLOCK);
     m_base = event_base_new_with_config(m_cfg);
     assert(m_base);
     if (m_base == NULL) {
-        throw voltdb::LibEventException();
+        throw LibEventException("Failed to create and initialize main event base");
     }
-    m_enableAbandon = config.m_enableAbandon;
-    m_hashScheme = config.m_hashScheme;
-    if (m_hashScheme == HASH_SHA1) {
-        SHA1_CTX context;
-        SHA1_Init(&context);
-        SHA1_Update( &context, reinterpret_cast<const unsigned char*>(config.m_password.data()), config.m_password.size());
-        m_passwordHash = (unsigned char *)malloc(20*sizeof(char));
-        SHA1_Final ( &context, m_passwordHash);
-    } else if (config.m_hashScheme == HASH_SHA256) {
-        m_passwordHash = (unsigned char *)malloc(32*sizeof(char));
-        computeSHA256(config.m_password.c_str(), config.m_password.size(), m_passwordHash);
-    } else {
-        throw voltdb::LibEventException();
+    hashPassword(config.m_password);
+    if (m_enableSSL) {
+        initSsl();
     }
-
     m_wakeupPipe[0] = -1;
     m_wakeupPipe[1] = -1;
 
-    m_timerMonitorBase = event_base_new();
-    assert(m_timerMonitorBase);
-    if (m_timerMonitorBase == NULL) {
-        throw voltdb::LibEventException();
+    if (m_enableQueryTimeout) {
+        m_timerMonitorBase = event_base_new();
+        if (m_timerMonitorBase == NULL) {
+            throw LibEventException("Failed to create and initialize event base for query-timeout monitor");
+        }
     }
     m_timerCheckPipe[0] = -1;
     m_timerCheckPipe[1] = -1;
 
-    //std::cout << "Scan interval: " << m_scanIntervalForTimedoutQuery.tv_sec << ":" << m_scanIntervalForTimedoutQuery.tv_usec << std::endl;
-    //std::cout << "Expiration interval: " << m_queryExpirationTime.tv_sec << ":" << m_scanIntervalForTimedoutQuery.tv_usec << std::endl;
+    ++m_numberOfClients;
```

Comment:
Using this as reference counter to track the number of clients that the user has spawned on separate threads. Multiple clients running on it's own thread use common openssl (initlaized once) resources that needs to be released when all the clients are done.